app/PaymentDrivers/Blockonomics/Blockonomics.php

```diff
@@ -137,43 +139,52 @@ public function paymentResponse(PaymentResponseRequest $request)
             'btc_address' => ['required'],
             'btc_amount' => ['required'],
             'btc_price' => ['required'],
-            // Setting status to required will break the payment process
-            // because sometimes the status is returned as 0 which is falsy
-            // and the validation will fail.
-            // 'status' => ['required'],
         ]);
 
+        $this->payment_hash = PaymentHash::where('hash', $request->payment_hash)->firstOrFail();
+
+        // Calculate fiat amount from Bitcoin
+        $amount_received_satoshis = $request->btc_amount;
+        $amount_satoshis_in_one_btc = 100000000;
+        $amount_received_btc = $amount_received_satoshis / $amount_satoshis_in_one_btc;
+        $price_per_btc_in_fiat = $request->btc_price;
+        $fiat_amount = round(($price_per_btc_in_fiat * $amount_received_btc), 2);
+
+        // Get the expected amount from payment hash
+        $payment_hash_data = $this->payment_hash->data;
+        $expected_amount = $payment_hash_data->amount_with_fee;
+
+        // ALWAYS adjust invoice allocations to match actual received amount
+        if ($fiat_amount != $expected_amount) {
+            $this->adjustInvoiceAllocations($fiat_amount);
+
+            // CRITICAL: Reload both payment_hash references
+            $this->payment_hash = PaymentHash::where('hash', $request->payment_hash)->firstOrFail();
```

Comment:
Is this duplicated for a reason? Or is this to handle the case where the payment hash data values have change?

If they have, it would be work including this in the if() block above

Reply:
Ah, it was not needed. Now it's been removed, good catch!